### PR TITLE
Blocks: Introduce helper function to retrieve hooked blocks

### PR DESCRIPTION
**Pull request by gziolo**
Trac ticket: https://core.trac.wordpress.org/ticket/59383

In order to implement Block Hooks (see [#59313](https://core.trac.wordpress.org/ticket/59313)), we added `block_hooks` field to the `WP_Block_Type` class, as well as to block registration related functions. In this follow-up, new helper function gets introduced that is going to compute the list of hooked blocks by other registered blocks for a given block type.

Extracted from https://github.com/WordPress/wordpress-develop/pull/5158 and covered with unit tests.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/blocks/blockHooks.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+					'tests/hooked-first-child' => 'first_child',
+					'tests/hooked-last-child'  => 'last_child',
+				),
+			)
+		);
+
+		$this->assertSame(
+			$expected,
+			get_hooked_blocks( $block_name )
+		);
```
Comment (ockham) [inline tests/phpunit/tests/blocks/blockHooks.php]: TBH I'm not sure a data provider is quite the right fit here. The way I understand it, a data provider makes sense to de-duplicate both setup and assertions across a set of different values; but in our case, the setup (hooked block registration) is always the same, and all of it is needed for all assertions. So I'd prefer if we inlined all assertions here in the test 😊

**Comment by ockham** [comment]
Looking good overall, just one [request](https://github.com/WordPress/wordpress-develop/pull/5239#pullrequestreview-1630643555) regarding tests.

For the other PRs we've filed for block hooks, I've actually filed separate tickets; it'd be great if we could do the same here 😊 (I can do that if you're busy with other things.)

Finally, just noting that I agree that we should eventually cache `get_hooked_blocks` as we [discussed](https://github.com/WordPress/wordpress-develop/pull/5158#discussion_r1326013492); but let's keep that for a follow-up.

**Comment by gziolo** [comment]
> For the other PRs we've filed for block hooks, I've actually filed separate tickets; it'd be great if we could do the same here 😊 (I can do that if you're busy with other things.)

Feel free to open another ticket. Otherwise, I'm fine referencing the existing ticket that covers the block hooks feature.

File: tests/phpunit/tests/blocks/blockHooks.php (hunk 1 above)
Comment (gziolo) [reply tests/phpunit/tests/blocks/blockHooks.php]: I had it like it before, I guess it's fine the way you described it with assertion messages. I will change it.

**Comment by gziolo** [comment]
I addressed the feedback, and I will open a ticket for this functionality now that I have some additional time.

**Comment by gziolo** [comment]
The ticket is https://core.trac.wordpress.org/ticket/59383.

File: tests/phpunit/tests/blocks/blockHooks.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			if ( $registry->is_registered( $block_name ) ) {
+				$registry->unregister( $block_name );
+			}
+		}
+
+		parent::tear_down();
+	}
+
+	/**
+	 * @ticket 59313
```
Comment (ockham) [inline tests/phpunit/tests/blocks/blockHooks.php]:
```suggestion
	 * @ticket 59383
```

File: tests/phpunit/tests/blocks/blockHooks.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	 * @covers ::get_hooked_blocks
+	 */
+	public function test_get_hooked_blocks_no_match_found() {
+		$result = get_hooked_blocks( 'tests/no-hooked-blocks' );
+
+		$this->assertSame( array(), $result );
+	}
+
+	/**
+	 * @ticket 59313
```
Comment (ockham) [inline tests/phpunit/tests/blocks/blockHooks.php]:
```suggestion
	 * @ticket 59383
```

**Review by ockham — COMMENTED** [review]
Thank you very much! Let's update the two `@ticket` references in the tests to use the new number, and it's ready to go 🎉

File: tests/phpunit/tests/blocks/blockHooks.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+			'block hooked at the first child position'
+		);
+		$this->assertSame(
+			array(
+				'tests/my-container-block' => 'last_child',
+			),
+			get_hooked_blocks( 'tests/hooked-last-child' ),
+			'block hooked at the last child position'
+		);
+	}
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/blockHooks.php]:
```suggestion
	public function test_get_hooked_blocks_matches_found() {
		register_block_type(
			'tests/injected-one',
			array(
				'block_hooks' => array(
					'tests/hooked-at-before' => 'before',
					'tests/hooked-at-after'  => 'after',
				),
			)
		);
		register_block_type(
			'tests/injected-two',
			array(
				'block_hooks' => array(
					'tests/hooked-at-before'      => 'before',
					'tests/hooked-at-after'       => 'after',
					'tests/hooked-at-first-child' => 'first_child',
					'tests/hooked-at-last-child'  => 'last_child',
				),
			)
		);

		$this->assertSame(
			array(
				'tests/injected-one' => 'before',
				'tests/injected-two' => 'before',
			),
			get_hooked_blocks( 'tests/hooked-at-before' ),
			'block hooked at the before position'
		);
		$this->assertSame(
			array(
				'tests/injected-one' => 'after',
				'tests/injected-two' => 'after',
			),
			get_hooked_blocks( 'tests/hooked-at-after' ),
			'block hooked at the after position'
		);
		$this->assertSame(
			array(
				'tests/injected-two' => 'first_child',
			),
			get_hooked_blocks( 'tests/hooked-at-first-child' ),
			'block hooked at the first child position'
		);
		$this->assertSame(
			array(
				'tests/injected-two' => 'last_child',
			),
			get_hooked_blocks( 'tests/hooked-at-last-child' ),
			'block hooked at the last child position'
		);
	}
```

**Comment by gziolo** [comment]
Committed with https://core.trac.wordpress.org/changeset/56610.

**Pull request closed without merging**